### PR TITLE
EBNT-269 EBNT-270 Pre-build images with prod-ready Flask server (uwsgi, nginx)

### PR DESCRIPTION
**Pull request by i-a-andrianov**

### CODE REVIEW

File: src/ebonite/ext/flask/prebuild/Dockerfile.j2 [hunk 1]
```diff
@@ -0,0 +1,25 @@
...
+WORKDIR /app
+
+RUN true \
+    && apt update \
+    && apt install -y \
+        bash \
+        nginx \
+        supervisor \
+        gcc \
+    && pip install uwsgi==2.0.18 \
```
Comment (mike0sv) [inline src/ebonite/ext/flask/prebuild/Dockerfile.j2:12]: bump version?

File: tests/build/test_docker.py [hunk 2]
```diff
@@ -0,0 +1,35 @@
...
+    assert image_exists_at_dockerhub('postgres:alpine')
+    assert image_exists_at_dockerhub('registry:latest')
+
+
+@pytest.mark.docker
+@pytest.mark.skipif(not has_docker(), reason='no docker installed')
+def test_image_not_exists():
+    assert not image_exists_at_dockerhub(f'python:100.200.300')
+    assert not image_exists_at_dockerhub('ebonite:forever')
+    assert not image_exists_at_dockerhub('minio:minio')
```
Comment (mike0sv) [inline tests/build/test_docker.py:22]: it's not clear that those images really do not exist))) i cases like this i prefer to use names like "minio:this_tag_definitely_does_not_exist_ololo"

File: src/ebonite/ext/prebuild.py [hunk 3]
```diff
@@ -0,0 +1,62 @@
...
+
+    with tempfile.TemporaryDirectory() as tmpdir:
+        context_dir = os.path.join(tmpdir, 'context')
+        logger.info('Building image %s on %s ...', tag, context_dir)
+
+        shutil.copytree(prebuild_path, context_dir)
+        _generate_dockerfile(context_dir, python_version)
+
+        try:
+            with create_docker_client() as client:
```
Comment (mike0sv) [inline src/ebonite/ext/prebuild.py:28]: this looks a lot like docker_builder code. cant we reuse it by implementing ServerProvider? We can then reuse it for different servers

File: src/ebonite/ext/prebuild.py [hunk 4]
```diff
@@ -0,0 +1,62 @@
...
+                client.images.build(path=context_dir, tag=tag, rm=True)
+
+            logger.info('Successfully built image %s', tag)
+        except Exception as e:
+            logger.error('Failed to build image %s: %s', tag, e)
+            return
+
+    if push:
+        with create_docker_client() as client:
+            login_to_registry(client, RemoteDockerRegistry())
```
Comment (mike0sv) [inline src/ebonite/ext/prebuild.py:38]: this requires setting env variables?

File: src/ebonite/ext/prebuild.py [hunk 5]
```diff
@@ -0,0 +1,62 @@
...
+
+
+def _generate_dockerfile(context_dir, python_version):
+    j2 = Environment(loader=FileSystemLoader([context_dir]))
+    docker_tmpl = j2.get_template('Dockerfile.j2')
+    with open(os.path.join(context_dir, 'Dockerfile'), 'w') as df:
+        df.write(docker_tmpl.render({'python_version': python_version}))
+
+
+PY_BLACK_LIST = {'^2\\.', '^3\\.[0-5]\\.', '^3\\.6\\.[0-4]'}
```
Comment (mike0sv) [inline src/ebonite/ext/prebuild.py:49]: what is up with 3.6.0-3.6.4?

File: src/ebonite/build/docker.py [hunk 6]
```diff
@@ -44,11 +42,8 @@ def get_host(self) -> str:
 class RemoteDockerRegistry(DockerRegistry):
     type = 'remote'
 
-    def __init__(self, host: str):
-        if re.match(VALID_HOST_REGEX, host):
-            self.host = host
-        else:
-            raise ValueError('Host {} is not valid'.format(host))
+    def __init__(self, host: Optional[str] = None):
+        self.host = host or 'https://index.docker.io/v1/'
```
Comment (mike0sv) [inline src/ebonite/build/docker.py:46]: why is this needed?

File: src/ebonite/ext/flask/build/post_install.j2 [hunk 7]
```diff
@@ -0,0 +1 @@
+ENV EBONITE_RUN_FLASK=false
```
Comment (mike0sv) [inline src/ebonite/ext/flask/build/post_install.j2:1]: why not in Dockerfile.j2?

File: src/ebonite/ext/flask/prebuild/Dockerfile.j2 [hunk 8]
```diff
@@ -0,0 +1,25 @@
...
+    && apt remove gcc -y \
+    && apt autoremove -y \
+    && apt clean
+
+RUN pip install flask==1.1.2 flasgger==0.9.3
```
Comment (mike0sv) [inline src/ebonite/ext/flask/prebuild/Dockerfile.j2:17]: maybe template it (at least flask) and use the current version in your system (to be consistent with things like run_model_server)

File: src/ebonite/ext/flask/server.py [hunk 9]
```diff
@@ -27,6 +22,7 @@ def create_executor_function(interface: Interface, method: str):
     :param method: method name
     :return: callable view function
     """
+    from flask import g, jsonify, request, send_file
```
Comment (mike0sv) [inline src/ebonite/ext/flask/server.py:25]: why localize imports?

File: src/ebonite/ext/flask/server.py [hunk 10]
```diff
@@ -75,31 +75,32 @@ class FlaskConfig(Config):
     FlaskConfig.log_params()
 
 
+PREBUILD_PATH = current_module_path('prebuild')
+BASE_IMAGE_TEMPLATE = 'zyfraai/flask:{}'
```
Comment (mike0sv) [inline src/ebonite/ext/flask/server.py:79]: i propose clearer name smth like `ebonite-flask-server`

File: src/ebonite/ext/flask/server.py [hunk 11]
```diff
@@ -75,31 +75,32 @@ class FlaskConfig(Config):
...
     ]
 
     # TODO somehow make this depend on builder implementation?
     additional_options = {
-        'templates_dir': os.path.join(os.path.dirname(__file__), TEMPLATES_DIR),
-        'run_cmd': '["/usr/bin/supervisord"]'
+        'templates_dir': current_module_path('build'),
+        'base_image': lambda python_version: BASE_IMAGE_TEMPLATE.format(python_version),
+        'run_cmd': False,  # base image has already specified command
+        'prebuild_hook': lambda python_version: prebuild_image(PREBUILD_PATH, BASE_IMAGE_TEMPLATE, python_version)
```
Comment (mike0sv) [inline src/ebonite/ext/flask/server.py:96]: maybe move it to separate class field?

File: src/ebonite/ext/prebuild.py [hunk 12]
```diff
@@ -0,0 +1,62 @@
...
+from jinja2 import Environment, FileSystemLoader
+
+from ebonite.build.docker import (RemoteDockerRegistry, create_docker_client, image_exists_at_dockerhub,
+                                  login_to_registry, repository_tags_at_dockerhub)
+from ebonite.utils.log import logger
+
+
+def prebuild_image(prebuild_path, name_template, python_version, *, push=False):
+    tag = name_template.format(python_version)
+    if image_exists_at_dockerhub(tag):
```
Comment (mike0sv) [inline src/ebonite/ext/prebuild.py:16]: maybe pull it here?

File: src/ebonite/ext/prebuild.py [hunk 13]
```diff
@@ -0,0 +1,62 @@
...
+    j2 = Environment(loader=FileSystemLoader([context_dir]))
+    docker_tmpl = j2.get_template('Dockerfile.j2')
+    with open(os.path.join(context_dir, 'Dockerfile'), 'w') as df:
+        df.write(docker_tmpl.render({'python_version': python_version}))
+
+
+PY_BLACK_LIST = {'^2\\.', '^3\\.[0-5]\\.', '^3\\.6\\.[0-4]'}
+
+
+def prebuild_missing_images(prebuild_path, name_template):
```
Comment (mike0sv) [inline src/ebonite/ext/prebuild.py:52]: this code and it's helper functions feel like they dont belong to ebonite codebase =\ anyway there's gonna be something that is gonna run it on schedule, maybe move it there completely?

File: src/ebonite/ext/flask/prebuild/Dockerfile.j2 (hunk 1 above)
Comment (i-a-andrianov) [reply src/ebonite/ext/flask/prebuild/Dockerfile.j2:12]: uWSGI 2.0.18 is the latest version in PyPi. What do you mean by "bump version"?

File: tests/build/test_docker.py (hunk 2 above)
Comment (i-a-andrianov) [reply tests/build/test_docker.py:22]: Do not see any difference here between your and my examples:
1. Python: it is at version 3 after 20 years, so version 100 is unrealistic
1. Minio: it seems unreasonable to use package name as its version, thus unrealistic
1. Ebonite: There is no such images at all and we fully control this

File: src/ebonite/ext/prebuild.py (hunk 3 above)
Comment (i-a-andrianov) [reply src/ebonite/ext/prebuild.py:28]: 1. Obviously we can reuse builder code but the main point was to detach this script from Ebonite completely in future.
1. I expect this script to be reused for all servers. That's why I put it there and not in `flask` package.

File: src/ebonite/ext/prebuild.py (hunk 4 above)
Comment (i-a-andrianov) [reply src/ebonite/ext/prebuild.py:38]: Yes. Variable names are composed according to rules declared by this method.

File: src/ebonite/ext/prebuild.py (hunk 5 above)
Comment (i-a-andrianov) [reply src/ebonite/ext/prebuild.py:49]: Unfortunately the latest uwsgi could not be installed for these versions of Python. That looks strange for me as it works fine for some (not all) 3.5.x versions but that's how it is. I think it should not be a problem as we should start to support 3.8 soon and give up on 3.6 concurrently.

File: src/ebonite/build/docker.py (hunk 6 above)
Comment (i-a-andrianov) [reply src/ebonite/build/docker.py:46]: This is an address of DockerHub's registry. I think it's good to specify it as default here not to require our users to Google for it (strangely enough it is not so easy to find).

File: src/ebonite/ext/flask/build/post_install.j2 (hunk 7 above)
Comment (i-a-andrianov) [reply src/ebonite/ext/flask/build/post_install.j2:1]: Do you mean `Dockerfile.j2` from Docker builder or from prebuild? For the former we could not specify this as builder uses generic template for all servers/models. For the latter I decided to make prebuilt images fully Ebonite-agnostic (name and internals) to promote their usage in other contexts. For DockerHub users they should look like alternative to FastAPI images prebuilt by FastAPI author.

File: src/ebonite/ext/flask/prebuild/Dockerfile.j2 (hunk 8 above)
Comment (i-a-andrianov) [reply src/ebonite/ext/flask/prebuild/Dockerfile.j2:17]: My point was that we should not require user to install server packages in their Jupyter notebooks. Moreover, it is simpler to control when versions are specified by us and not by user environment. For ML libraries it is unrealistic (it could break user's models) but for servers it looks feasible and useful (server-related code is written by us and could be on the contrary broken by user-specified version).

Comment (i-a-andrianov) [reply src/ebonite/ext/flask/prebuild/Dockerfile.j2:17]: As for run_model_server I think it's OK not to be 100% consistent with it as it is a bit hackish by itself. For me it's more like a workaround to overcome slowness of images building (which was greatly reduced by this PR but we should work on it further).

File: src/ebonite/ext/flask/server.py (hunk 9 above)
Comment (i-a-andrianov) [reply src/ebonite/ext/flask/server.py:25]: To require Flask/flasgger installation inside runtime (images) only and not during work with metadata (in Jupyter notebooks, etc). This is similar to sklearn model wrapper requiring sklearn installation when model is actually called only.

File: src/ebonite/ext/flask/server.py (hunk 10 above)
Comment (i-a-andrianov) [reply src/ebonite/ext/flask/server.py:79]: I decided to make prebuilt images fully Ebonite-agnostic (name and internals) to promote their usage in other contexts. For DockerHub users they should look like alternative to FastAPI images prebuilt by FastAPI author.

File: src/ebonite/ext/flask/server.py (hunk 11 above)
Comment (i-a-andrianov) [reply src/ebonite/ext/flask/server.py:96]: Do you mean move `prebuild_hook` to class field? Of which class? Please provide a bit more details on that. I did it here as existing options here were already used for Docker builder configuration. Thus I though it was a good fit.

File: src/ebonite/ext/prebuild.py (hunk 12 above)
Comment (i-a-andrianov) [reply src/ebonite/ext/prebuild.py:16]: Do you mean pulling `tag` image locally?

I think it is a bad idea. Running this script (as main) would currently pull you ~20 large images (Python+Flask over it) even if you won't do anything with them. That's why I didn't use Docker client here but called Dockerhub endpoint directly. With current solution even after `docker image prune` command running this script takes several seconds, leaves no traces in Docker on my machine and just says that all images were already built.

When this code is used inside Docker builder it's OK as it Docker would pull 1 image itself (it is caused by `FROM` statement in model image `Dockerfile`).

File: src/ebonite/ext/prebuild.py (hunk 13 above)
Comment (i-a-andrianov) [reply src/ebonite/ext/prebuild.py:52]: Yep, I planned to detach this code completely from Ebonite in future. But we should be aware of potentially high maintenance burden (dealing with several repos) in this case. That's why for now I decided to leave it here but already make it as Ebonite-agnostic as possible.

File: src/ebonite/ext/flask/prebuild/Dockerfile.j2 (hunk 1 above)
Comment (mike0sv) [reply src/ebonite/ext/flask/prebuild/Dockerfile.j2:12]: ok, just thought that 2.0.18 refers to the release year (now that im writing it it actually sounds stupid)

File: tests/build/test_docker.py (hunk 2 above)
Comment (mike0sv) [reply tests/build/test_docker.py:22]: the difference is not that this could be real images, the difference is that when i see "minio:minio", i need to think about it, and when i see "minio:this_does_not_exist" i do not

File: src/ebonite/ext/prebuild.py (hunk 3 above)
Comment (mike0sv) [reply src/ebonite/ext/prebuild.py:28]: but it still will depend on ebonite, so it can use it internals. but ok

File: src/ebonite/build/docker.py (hunk 6 above)
Comment (mike0sv) [reply src/ebonite/build/docker.py:46]: yes, but i thought that it was default for docker - if you dont set it, docker registry is used, no?

File: src/ebonite/ext/flask/server.py (hunk 9 above)
Comment (mike0sv) [reply src/ebonite/ext/flask/server.py:25]: you mean to be able to build images with flask server even if you dont have flask installed?

Comment (mike0sv) [reply src/ebonite/ext/flask/server.py:25]: ok for now, but maybe we can do better here

File: src/ebonite/ext/flask/server.py (hunk 10 above)
Comment (mike0sv) [reply src/ebonite/ext/flask/server.py:79]: ok, but still just "flask" looks like you just installed flask over python image. at least "flask-server"? also, to promote use of it in other context, need some documentation (like that you need app.py file or something?)

File: src/ebonite/ext/flask/server.py (hunk 11 above)
Comment (mike0sv) [reply src/ebonite/ext/flask/server.py:96]: i think i meant Server class, not sure now if this will work.
but actually, the idea behind this comment is more general.
i like this idea of separating build into 2 stages - server and model. even if we do not prebuilt servers, still it allows to reuse build cache for example.
so it is natural to make this more "legit", because right now it looks like we hack our way around our own design, using "holes" we left to provide users a way to customize stuff.
so this proposition was one step towards this "legitimization", but maybe it is better to use complex approach and redisign this altogether

File: src/ebonite/ext/prebuild.py (hunk 12 above)
Comment (mike0sv) [reply src/ebonite/ext/prebuild.py:16]: my thought that side-effect of this function was "we have base server image" locally, either by building or pulling it.

File: src/ebonite/ext/flask/server.py (hunk 9 above)
Comment (i-a-andrianov) [reply src/ebonite/ext/flask/server.py:25]: Yes, I mean building Flask images without Flask installation. I agree that imports are kind of dumb solution. To do better than this we should decompose server runtime (how to expose model endpoints) and metadata (how to build image with server) parts for all servers. I decided this to be quite a cruel change in scope of this PR. But this could certainly be taken into account in future.

File: src/ebonite/build/docker.py (hunk 6 above)
Comment (i-a-andrianov) [reply src/ebonite/build/docker.py:46]: In earlier version of this class we were required to specify host. Simply putting `None` caused and still causes exceptions in several parts of our code (at least, `DockerImage.get_uri` and `login_to_registry`). Moreover, Docker client `login` method is kind of cryptic here: it has `registry=None` but its documentation says "registry (str): URL to the registry.  E.g. ``https://index.docker.io/v1/``". Considering all of this I decided to specify this URL as a safety net.

File: tests/build/test_docker.py (hunk 2 above)
Comment (i-a-andrianov) [reply tests/build/test_docker.py]: ok, I've got your point, fixed

File: src/ebonite/ext/flask/server.py (hunk 10 above)
Comment (i-a-andrianov) [reply src/ebonite/ext/flask/server.py:79]: For me it's ok not to use word "server" here. Moreover, it's somewhat consistent with "uvicorn-gunicorn-fastapi". It would be even more consistent if we have "uwsgi-nginx-flask" but "uwsgi-nginx" are implementation details and can be omitted.

For me the lack of documentation actually seems like much more important problem here. You are right, `app.py` should be replaced in order to use this image. I need to take some time to prepare documentation for the repo on Dockerhub. But I don't think it should block this PR somehow. I will notify you privately when docs are ready.
